src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -910,24 +912,69 @@ private void ScanSyntaxToken(ref TokenInfo info)
+                        // skip the pair in its entirety so we can produce a good error message that treats the two as a
+                        // single entity to tell the user about.
+                        if (char.IsHighSurrogate(character))
+                        {
+                            if (char.IsLowSurrogate(TextWindow.PeekChar()))
+                                TextWindow.AdvanceChar();
+                        }
                     }
 
+                    // if the original text wasn't already escaped, then escape it in the error message so that it's
+                    // clear what the issue is.  i.e. if the users source had the literal six characters in order like
+                    // so:  `\` `u` `D` `C` `E` `7`, then there's no need to escape that again when reporting the error.
+                    // However, if the user's code has the actual System.Char \uDCE7 char in it, then we want to print
+                    // that out in escaped form so they have an actual clue about what the character value is that we
+                    // have a problem with.
                     if (_badTokenCount++ > 200)
                     {
                         // If we get too many characters that we cannot make sense of, absorb the rest of the input.
                         int end = TextWindow.Text.Length;
-                        int width = end - startingPosition;
-                        info.Text = TextWindow.Text.ToString(new TextSpan(startingPosition, width));
+                        info.Text = TextWindow.Text.ToString(TextSpan.FromBounds(startingPosition, end));
                         TextWindow.Reset(end);
                     }
                     else
                     {
                         info.Text = TextWindow.GetText(intern: true);
                     }
 
-                    this.AddError(ErrorCode.ERR_UnexpectedCharacter, info.Text);
+                    // if the original user code already contained an explicit escape-sequence in text, then we can just
+                    // point to that code exactly as being something we didn't expect.  Otherwise, if it's just normal
+                    // characters, then escape anything non-printable so that the error message clearly indicates what
+                    // the char-values were that we ran into.
+                    var messageText = isEscaped ? info.Text : escapeNonPrintableCharacters(info.Text);
+                    this.AddError(ErrorCode.ERR_UnexpectedCharacter, messageText);
                     break;
             }
+
+            static string escapeNonPrintableCharacters(string str)
+            {
+                var sb = PooledStringBuilder.GetInstance();
+                for (int i = 0, n = str.Length; i < n; i++)
+                {
+                    var c = str[i];
+
+                    if (i < n - 1 &&
+                        char.IsHighSurrogate(c) &&
+                        char.IsLowSurrogate(str[i + 1]))
```

Comment:
>IsLowSurrogate [](http://example.com/codeflow?start=29&length=14)

Nit: wouldn't this work? `char.IsSurrogatePair(str, i)`

Reply:
yup.  it would.  thanks :)